### PR TITLE
feat(liquidity mining): liquidity rolling logic

### LINKED ISSUE #1871: Liquidity mining distribution script should handle multiple pools
**Why**

To execute rolls, the liquidity mining script will need to be able to take into account liquidity in multiple pools at once.

**How**

The script should distribute the $UMA based on an address' yUSD in both pools relative to the total yUSD across both pools.

**Resolution**

Self-explanatory.

**Difficulty Score [1-10]**

6: ~1 day

### PR DESCRIPTION
**Pull request by chrismaree**
**Motivation**

issue: #1871

Adding rolling between different balancer pools support.

**Details**

There are number of ways to implement this feature. For example, the existing `CalculateBalancerLPRewards.js` script could be modified to take in multiple pools. This seems to be the most reasonable implementation. However, major complexity is introduced when you need to think about how the timing and division of rewards occur between the roll part of the week (two pools) and the non-roll part of the week (one pool, like the normal setup).

As a result, this PR archives the rolling functionality by leaving the existing `CalculateBalancerLPRewards.js` script as it is to calculate rewards for the single pool case and then introduces a second script `CalculateRollBalancerLPRewards.js` which considers multiple pool inputs in its reward calculation. Then, a third script `JoinRolledPayouts.js` is added which is used to combined the payouts of script 1 & 2 to get the net weekly payout.

While having 3 scripts seems to be overly complex, what it does is keep each script's domain smaller and easier to reason about. We can then combined the scripts in different configurations depending on the LM program we are running.

**The scripts would be used as follows:**
For the last week, we ran rewards split over both pools from Aug 24 until Aug 28. Then, Aug 28 until Aug 31 we considered only the "new" Oct pool. Using these scripts, we would run the `CalculateRollBalancerLPRewards.js` between `Aug 24 until Aug 28` with `rollFrac` fraction of the rewards and then run `CalculateBalancerLPRewards.js` with the remaining rewards. This fraction of rewards would be calculated by looking at the block time duration that covers both pools relative to the total rewards week duration:

```
rollFrac = (roll_end_block - rewards_start_block) / (rewards_end_block - rewards_start_block)
```

remaining weekly rewards =  `1 - rollFrac`

Then, the rewards output would be joined together using the `JoinRolledPayouts.js`.

In the process, the `CalculateRollBalancerLPRewards.js` script creates an output file for the roll rewards that considers both pools. The output from `CalculateRollBalancerLPRewards.js` is also preserved to enable validation of the two phases within the rewards weekly payout.

**This PR introduces the following additional changes:**

1) Small refactor `CalculateBalancerLPRewards.js` to include more configuration options, such as `umaPerWeek`, `blocksPerSnapshot` as well.
2) Addition of extra meta data to previous payouts such as `umaPerWeek` `blocksPerSnapshot` and `poolAddress`. These become important when we consider the rolling outputs.

Lastly, I tried to move the tests into a different directory (`core/tests/scripts/liquidity-mining`) but had issues with buidler and `advanceBlockAndSetTime` function calls. I spent 2 hours trying to fix this but couldn't get it to pass correctly. For now, we can leave this outside of CI until there is a fix to this issue.

**Issue(s)**

Fixes #1871

### CODE REVIEW

File: packages/core/scripts/liquidity-mining/CalculateRollBalancerLPRewards.js [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+  console.log("⚖️  Finding balancer pool info...");
+  // Get the information on a particular pool. This includes a mapping of all previous token holders (shareholders).
+  const pool1Info = await _fetchBalancerPoolInfo(pool1Address);
+  const pool2Info = await _fetchBalancerPoolInfo(pool2Address);
+
+  // Extract the addresses of all historic shareholders.
+  const pool1ShareHolders = pool1Info.shares.flatMap(a => a.userAddress.id);
+  const pool2ShareHolders = pool2Info.shares.flatMap(a => a.userAddress.id);
+  const shareHolders = [...pool1ShareHolders, ...pool2ShareHolders];
```
Comment (chrismaree) [inline packages/core/scripts/liquidity-mining/CalculateRollBalancerLPRewards.js:68]: core change in this script 1) shareholders considers both pools.

File: packages/core/scripts/liquidity-mining/CalculateRollBalancerLPRewards.js [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+  }
+  const balanceResultsbPool1 = await Promise.allSettled(promiseArraybPool1);
+  await delay(5); // slow down the queries between these massive promise arrays. Helps to keep Infura happy.
+  const balanceResultsbPool2 = await Promise.allSettled(promiseArraybPool2);
+  // For each balance result, calculate their associated payment addition.
+  Object.entries(shareHolderPayout).forEach(function(shareHolder, index) {
+    // If the given shareholder had no BLP tokens at the given block, skip them.
+    if (balanceResultsbPool1[index].value === "0" && balanceResultsbPool2[index].value === "0") return;
+    // The holders fraction is the number of BPTs at the block divided by the total supply at that block.
+    const shareHolderBalanceAtSnapshot = toBN(balanceResultsbPool1[index].value).add(
```
Comment (chrismaree) [inline packages/core/scripts/liquidity-mining/CalculateRollBalancerLPRewards.js:158]: main change 3) shareholder balance is the sum over both pools

File: packages/core/scripts/liquidity-mining/CalculateRollBalancerLPRewards.js [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+  progressBar.stop();
+
+  return shareHolderPayout;
+}
+
+// For a given `blockNumber` (snapshot in time), return an updated `shareHolderPayout` object that has appended
+// payouts for a given `bPool` at a rate of `umaPerSnapshot`.
+async function _updatePayoutAtBlock(bPool1, bPool2, blockNumber, shareHolderPayout, umaPerSnapshot) {
+  // Get the total supply of Balancer Pool tokens at the given snapshot's block number.
+  const bptSupplyAtSnapshot = toBN(await bPool1.methods.totalSupply().call(undefined, blockNumber)).add(
```
Comment (chrismaree) [inline packages/core/scripts/liquidity-mining/CalculateRollBalancerLPRewards.js:139]: main change 2) total supply considers both balancer pools

File: packages/core/scripts/liquidity-mining/tests/CalculateBalancerLPRewards.js [hunk 4]
```diff
@@ -2,8 +2,7 @@ const { toWei, toBN } = web3.utils;
 
 const { advanceBlockAndSetTime } = require("@umaprotocol/common");
 
-const Main = require("../CalculateBalancerLPProviders"); // Script to test.
-const { _updatePayoutAtBlock } = require("../CalculateBalancerLPProviders");
+const { _updatePayoutAtBlock, _calculatePayoutsBetweenBlocks } = require("../CalculateBalancerLPRewards");
```
Comment (chrismaree) [inline packages/core/scripts/liquidity-mining/tests/CalculateBalancerLPRewards.js:5]: small clean

File: packages/core/scripts/liquidity-mining/CalculateBalancerLPRewards.js [hunk 5]
```diff
@@ -20,29 +20,32 @@ const web3 = new Web3(new Web3.providers.HttpProvider(process.env.CUSTOM_NODE_UR
...
 const argv = require("minimist")(process.argv.slice(), {
-  string: ["poolAddress", "fromDate", "toDate"],
-  integer: ["week"],
-  boolean: ["test"]
+  string: ["poolAddress", "tokenName"],
+  integer: ["fromBlock", "toBlock", "week", "umaPerWeek", "blocksPerSnapshot"]
 });
 
-const UMA_PER_WEEK = toBN(toWei("25000"));
-const BLOCKS_PER_SNAPSHOT = 256;
```
Comment (nicholaspai) [inline packages/core/scripts/liquidity-mining/CalculateBalancerLPRewards.js:29]: Why not make this a constant anymore? This shouldn't be affected by rolling right?

Comment (chrismaree) [reply packages/core/scripts/liquidity-mining/CalculateBalancerLPRewards.js:29]: Right but it's useful for testing. It defaults to 256 in the input params.

File: packages/core/scripts/liquidity-mining/CalculateRollBalancerLPRewards.js [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+
+  // Get the given holders balance at the given block. Generate an array of promises to resolve in parallel.
+  let promiseArraybPool1 = [];
+  let promiseArraybPool2 = [];
+  for (shareHolder of Object.keys(shareHolderPayout)) {
+    promiseArraybPool1.push(bPool1.methods.balanceOf(shareHolder).call(undefined, blockNumber));
+    promiseArraybPool2.push(bPool2.methods.balanceOf(shareHolder).call(undefined, blockNumber));
+  }
+  const balanceResultsbPool1 = await Promise.allSettled(promiseArraybPool1);
+  await delay(5); // slow down the queries between these massive promise arrays. Helps to keep Infura happy.
```
Comment (mrice32) [inline packages/core/scripts/liquidity-mining/CalculateRollBalancerLPRewards.js:151]: nit: what does infura do if you don't?

Would bluebird promises be a better way to throttle the total number of outstanding requests?

File: packages/core/scripts/liquidity-mining/CalculateBalancerLPRewards.js [hunk 7]
```diff
@@ -4,7 +4,7 @@
 // provided by for each liquidity provider to the single whitelisted pool.
 // -> For each snapshot block, calculate the $UMA rewards to be received by each liquidity provider based on the target weekly distribution.
 
-// Example usage from core: truffle exec ./scripts/liquidity-mining/calculateBalancerLPProviders.js --network mainnet_mnemonic --poolAddress="0x0099447ef539718bba3c4d4d4b4491d307eedc53" --fromDate="2020-07-06" --toDate="2020-07-13" --week=1
+// Example usage from core: truffle exec ./scripts/liquidity-mining/CalculateBalancerLPRewards.js --network mainnet_mnemonic --poolAddress="0x0099447ef539718bba3c4d4d4b4491d307eedc53" --fromDate="2020-07-06" --toDate="2020-07-13" --week=1
```
Comment (mrice32) [inline packages/core/scripts/liquidity-mining/CalculateBalancerLPRewards.js:7]: nit: can these be run with node? If so, do you think that should be the suggested way to run them now?

File: packages/core/scripts/liquidity-mining/CalculateRollBalancerLPRewards.js [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+  }
+  progressBar.stop();
+
+  return shareHolderPayout;
+}
+
+// For a given `blockNumber` (snapshot in time), return an updated `shareHolderPayout` object that has appended
+// payouts for a given `bPool` at a rate of `umaPerSnapshot`.
+async function _updatePayoutAtBlock(bPool1, bPool2, blockNumber, shareHolderPayout, umaPerSnapshot) {
+  // Get the total supply of Balancer Pool tokens at the given snapshot's block number.
```
Comment (mrice32) [inline packages/core/scripts/liquidity-mining/CalculateRollBalancerLPRewards.js:138]: As discussed offline, we probably want to weight by the number of yUSD tokens that these balances represent.

Comment (nicholaspai) [reply packages/core/scripts/liquidity-mining/CalculateRollBalancerLPRewards.js:165]: Ah this is a great point.

**Comment by chrismaree** [comment]
After much discussion offline with @mrice32 and @nicholaspai this PR's implementation has been changed in a subtle way. 

Previously, the number of balancer pool tokens the token sponsor holds was used to define their contribution to each pool during the roll. The issue with this is that it assumes that 1 BPT in pool1 = 1 BPT in pool2. This assumption is _almost_ correct but it can differ by an amount defined by the price that the balancer pool is seeded at. With the current Oct and Sept contracts, this amounts to a ~5% error. As each UMA is pushing 25 USD right now this amounts to 31 250 USD error, which is way above any reasonable tolerance.

As a result, the PR was refactored to consider the amount of yUSD each token sponsor adds to the pool, rather than just their BPT. This was done using the following formula applied to each token sponsor:

```
myReward = totalUMAPayout * (myyUSDinBP1 + myyUSDinBP2) / (totalyUSDinBP1 + totalyUSDinBP2)
```

The values for `myyUSDinBP1` and `myyUSDinBP2` were found by calculating how many yUSD each pool holds and then dividing this by the total number of BPT for that pool. From this, you can find a "price" of BPT, in yUSD. This multiplied by the individuals' BPT contribution will give the total yUSD they contributed to the given pool. 

The rest of the logic in the script remained the same, including how the scripts are run and joined together.

File: packages/core/scripts/liquidity-mining/CalculateBalancerLPRewards.js (hunk 7 above)
Comment (mrice32) [reply packages/core/scripts/liquidity-mining/CalculateBalancerLPRewards.js:7]: bump

File: packages/core/scripts/liquidity-mining/CalculateBalancerLPRewards.js [hunk 9]
```diff
@@ -20,29 +20,32 @@ const web3 = new Web3(new Web3.providers.HttpProvider(process.env.CUSTOM_NODE_UR
...
+  week,
+  umaPerWeek = 25000,
+  blocksPerSnapshot = 256
+) {
   // Create two moment objects from the input string. Convert to UTC time zone. As no time is provided in the input
   // will parse to 12:00am UTC.
-  if (!web3.utils.isAddress(poolAddress) || !fromBlock || !toBlock || !week) {
-    throw "Missing or invalid parameter! Provide poolAddress, fromBlock, toBlock & week.";
+  if (!web3.utils.isAddress(poolAddress) || !fromBlock || !toBlock || !week || !tokenName) {
+    throw "Missing or invalid parameter! Provide poolAddress, fromBlock, toBlock, week & tokenName";
```
[the remark this answers is not in the document]
Comment (mrice32) [reply packages/core/scripts/liquidity-mining/CalculateBalancerLPRewards.js:39]: bump

File: packages/core/scripts/liquidity-mining/CalculateRollBalancerLPRewards.js (hunk 6 above)
Comment (mrice32) [reply packages/core/scripts/liquidity-mining/CalculateRollBalancerLPRewards.js:199]: bump.

File: packages/core/scripts/liquidity-mining/CalculateBalancerLPRewards.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (chrismaree) [reply packages/core/scripts/liquidity-mining/CalculateBalancerLPRewards.js:39]: There were 4 scripts that all had this same throw with no `Error`. I have updated them accordingly.

File: packages/core/scripts/liquidity-mining/CalculateRollBalancerLPRewards.js (hunk 6 above)
Comment (chrismaree) [reply packages/core/scripts/liquidity-mining/CalculateRollBalancerLPRewards.js:199]: I implemented bluebird promises but I found that this timed out more and was slower. For some reason, infura is ok with a large burst (~2000 calls in parallel) if you then back off for a bit.

File: packages/core/scripts/liquidity-mining/CalculateBalancerLPRewards.js (hunk 7 above)
Comment (chrismaree) [reply packages/core/scripts/liquidity-mining/CalculateBalancerLPRewards.js:7]: indeed. Updated.

**Pull request merged**